src/lib/comms/sol-lwm2m-security.c

```diff
@@ -0,0 +1,470 @@
+
+SOL_LOG_INTERNAL_DECLARE(_lwm2m_security_domain, "lwm2m-security");
+
+#ifdef DTLS
+
+struct sol_socket *sol_coap_server_get_socket(const struct sol_coap_server *server);
+
+static ssize_t
+get_psk_from_server_or_bs_server(const void *data, struct sol_str_slice id,
+    char *psk, size_t psk_len)
+{
+    const struct sol_lwm2m_security *ctx = data;
+    struct sol_lwm2m_server *lwm2m_server;
+    struct sol_lwm2m_bootstrap_server *lwm2m_bs_server;
+    struct sol_vector *known_psks;
+    struct sol_lwm2m_security_psk *stored_psk;
+    uint16_t i;
+
+    SOL_DBG("Looking for PSK with ID=%.*s", SOL_STR_SLICE_PRINT(id));
+
+    //If the caller expects a PSK_LEN less than 16 bytes
+    if (psk_len < SOL_DTLS_PSK_KEY_LEN)
+        return -ENOBUFS;
+
+    if (ctx->type == LWM2M_SERVER) {
+        SOL_DBG("ctx->type = LWM2M_SERVER");
+        lwm2m_server = ctx->entity;
+        known_psks = lwm2m_server->known_psks;
+    } else {
+        SOL_DBG("ctx->type = LWM2M_BOOTSTRAP_SERVER");
+        lwm2m_bs_server = ctx->entity;
+        known_psks = lwm2m_bs_server->known_psks;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (known_psks, stored_psk, i) {
+        if (sol_str_slice_eq(sol_str_slice_from_blob(stored_psk->id), id)) {
+            if (stored_psk->key->size != SOL_DTLS_PSK_KEY_LEN) {
+                SOL_WRN("The PSK '%.*s' is %zu-bytes long; expecting a %d-bytes long PSK",
+                    SOL_STR_SLICE_PRINT(sol_str_slice_from_blob(stored_psk->key)),
+                    stored_psk->key->size, SOL_DTLS_PSK_KEY_LEN);
+                return -EINVAL;
+            }
+
+            memcpy(psk, stored_psk->key->mem, stored_psk->key->size);
+            return (ssize_t)SOL_DTLS_PSK_KEY_LEN;
+        }
+    }
+
+    return -ENOENT;
+}
+
+static ssize_t
+get_psk_from_client(const void *data, struct sol_str_slice id,
+    char *psk, size_t psk_len)
+{
+    const struct sol_lwm2m_security *ctx = data;
+    struct sol_lwm2m_client *lwm2m_client = ctx->entity;
+    struct obj_ctx *obj_ctx;
+    struct obj_instance *instance;
+    struct sol_lwm2m_resource res[3] = { };
+    uint16_t i;
+    int r;
+
+    SOL_DBG("Looking for PSK with ID=%.*s", SOL_STR_SLICE_PRINT(id));
+
+    //If the caller expects a PSK_LEN less than 16 bytes
+    if (psk_len < SOL_DTLS_PSK_KEY_LEN)
+        return -ENOBUFS;
+
+    obj_ctx = find_object_ctx_by_id(lwm2m_client, SECURITY_OBJECT_ID);
+    if (!obj_ctx) {
+        SOL_WRN("LWM2M Security object not provided!");
+        return -ENOENT;
+    }
+
+    if (!obj_ctx->instances.len) {
+        SOL_WRN("There are no Security Server instances");
+        return -ENOENT;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, instance, i) {
+        r = read_resources(lwm2m_client, obj_ctx, instance, res, 3,
+            SECURITY_SECURITY_MODE,
+            SECURITY_PUBLIC_KEY_OR_IDENTITY,
+            SECURITY_SECRET_KEY);
+        SOL_INT_CHECK_GOTO(r, < 0, err_clear);
+
+        //If -ENOENT when reading SECURITY_SECRET_KEY or
```

Comment:
`-ENOENT` is never actually returned by `read_resources`, since it's a valid situation to have an optional non-supported resource in there (in which case it will just be filled with empty values by `read_resources` itself, see https://github.com/solettaproject/soletta/blob/master/src/lib/comms/sol-lwm2m-client.c#L1840).

So, the `SOL_INT_CHECK_GOTO` would not jump to `err_clear`, and that's why this check is needed.